<p style="text-align:center">
    <a href="https://skills.network" target="_blank">
    <img src="https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/assets/logos/SN_web_lightmode.png" width="200" alt="Skills Network Logo">
    </a>
</p>


# Exploring and pre-processing a dataset using Pandas 


Estimated time needed: **30** minutes
    

## Objectives

After completing this lab you will be able to:

* Explore the dataset
* Pre-process dataset as required (may be for visualization)


## Introduction

The aim of this lab is to provide you a refresher on the **Pandas** library, so that you can pre-process and anlyse the datasets before applying data visualization techniques on it. This lab will work as acrash course on *pandas*. if you are interested in learning more about the *pandas* library, detailed description and explanation of how to use it and how to clean, munge, and process data stored in a *pandas* dataframe are provided in our course [**Data Analysis with Python**](https://www.coursera.org/learn/data-analysis-with-python?specialization=ibm-data-analyst) and [**Python for Applied Data Science**](https://www.coursera.org/learn/python-for-applied-data-science-ai?specialization=ibm-data-analyst)

------------


## Table of Contents

<div class="alert alert-block alert-info" style="margin-top: 20px">

1. [Exploring Datasets with *pandas*](#0)<br>
2. [The Dataset: Immigration to Canada from 1980 to 2013](#1)<br>
3. [*pandas* Basics](#2) <br>
4. [*pandas* Intermediate: Indexing and Selection](#3) <br>
5. [*pandas* Filtering based on a criteria](#4)<br>
6. [*pandas* Sorting Values](#5)

</div>


# Exploring Datasets with *pandas* <a id="0"></a>

*pandas* is an essential data analysis toolkit for Python. From their [website](http://pandas.pydata.org/):
>*pandas* is a Python package providing fast, flexible, and expressive data structures designed to make working with “relational” or “labeled” data both easy and intuitive. It aims to be the fundamental high-level building block for doing practical, **real world** data analysis in Python.

The course heavily relies on *pandas* for data wrangling, analysis, and visualization. We encourage you to spend some time and familiarize yourself with the *pandas* API Reference: http://pandas.pydata.org/pandas-docs/stable/api.html.


## The Dataset: Immigration to Canada from 1980 to 2013 <a id="1"></a>


Dataset Source: [International migration flows to and from selected countries - The 2015 revision](https://www.un.org/development/desa/pd/data/international-migration-flows).

The dataset contains annual data on the flows of international immigrants as recorded by the countries of destination. The data presents both inflows and outflows according to the place of birth, citizenship or place of previous / next residence both for foreigners and nationals. The current version presents data pertaining to 45 countries.

In this lab, we will focus on the Canadian immigration data.

![Data Preview](https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-DV0101EN-SkillsNetwork/labs/Module%201/images/DataSnapshot.png)

 The Canada Immigration dataset can be fetched from <a href="https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-DV0101EN-SkillsNetwork/Data%20Files/Canada.xlsx">here</a>.

---


## *pandas* Basics<a id="2"></a>


The first thing we'll do is install **openpyxl** (formerly **xlrd**), a module that *pandas* requires to read Excel files.


In [10]:
!pip install --upgrade xlrd


In [11]:
!pip install openpyxl


Next, we'll do is import two key data analysis modules: *pandas* and *numpy*.


In [8]:
import numpy as np  # useful for many scientific computing in Python
import pandas as pd # primary data structure library

Let's download and import our primary Canadian Immigration dataset using *pandas*'s `read_excel()` method.


In [12]:
df_can = pd.read_excel(
    'https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-DV0101EN-SkillsNetwork/Data%20Files/Canada.xlsx',
    sheet_name='Canada by Citizenship',
    skiprows=range(20),
    skipfooter=2,
    engine='openpyxl'
)

print('Data read into a pandas dataframe!')


Data read into a pandas dataframe!


Let's view the top 5 rows of the dataset using the `head()` function.


In [13]:
df_can.head()
# tip: You can specify the number of rows you'd like to see as follows: df_can.head(10) 

Type    Coverage          OdName   AREA AreaName     REG  \
0  Immigrants  Foreigners     Afghanistan  935.0     Asia  5501.0   
1  Immigrants  Foreigners         Albania  908.0   Europe   925.0   
2  Immigrants  Foreigners         Algeria  903.0   Africa   912.0   
3  Immigrants  Foreigners  American Samoa  909.0  Oceania   957.0   
4  Immigrants  Foreigners         Andorra  908.0   Europe   925.0   

           RegName    DEV             DevName  1980  ...    2012    2013  \
0    Southern Asia  902.0  Developing regions  16.0  ...  2635.0  2004.0   
1  Southern Europe  901.0   Developed regions   1.0  ...   620.0   603.0   
2  Northern Africa  902.0  Developing regions  80.0  ...  3774.0  4331.0   
3        Polynesia  902.0  Developing regions   0.0  ...     0.0     0.0   
4  Southern Europe  901.0   Developed regions   0.0  ...     1.0     1.0   

   Unnamed: 43  Unnamed: 44  Unnamed: 45  Unnamed: 46  Unnamed: 47  \
0          NaN          NaN          NaN          NaN          NaN   
1          NaN          NaN          NaN          NaN          NaN   
2          NaN          NaN          NaN          NaN          NaN   
3          NaN          NaN          NaN          NaN          NaN   
4          NaN          NaN          NaN          NaN          NaN   

   Unnamed: 48  Unnamed: 49  Unnamed: 50  
0          NaN          NaN          NaN  
1          NaN          NaN          NaN  
2          NaN          NaN          NaN  
3          NaN          NaN          NaN  
4          NaN          NaN          NaN  

[5 rows x 51 columns]

We can also view the bottom 5 rows of the dataset using the `tail()` function.


In [14]:
df_can.tail()

Type Coverage OdName  AREA AreaName  REG RegName  DEV DevName  1980  ...  \
1008  NaN      NaN    NaN   NaN      NaN  NaN     NaN  NaN     NaN   NaN  ...   
1009  NaN      NaN    NaN   NaN      NaN  NaN     NaN  NaN     NaN   NaN  ...   
1010  NaN      NaN    NaN   NaN      NaN  NaN     NaN  NaN     NaN   NaN  ...   
1011  NaN      NaN    NaN   NaN      NaN  NaN     NaN  NaN     NaN   NaN  ...   
1012  NaN      NaN    NaN   NaN      NaN  NaN     NaN  NaN     NaN   NaN  ...   

      2012  2013  Unnamed: 43  Unnamed: 44  Unnamed: 45  Unnamed: 46  \
1008   NaN   NaN          NaN          NaN          NaN          NaN   
1009   NaN   NaN          NaN          NaN          NaN          NaN   
1010   NaN   NaN          NaN          NaN          NaN          NaN   
1011   NaN   NaN          NaN          NaN          NaN          NaN   
1012   NaN   NaN          NaN          NaN          NaN          NaN   

      Unnamed: 47  Unnamed: 48  Unnamed: 49  Unnamed: 50  
1008          NaN          NaN          NaN          NaN  
1009          NaN          NaN          NaN          NaN  
1010          NaN          NaN          NaN          NaN  
1011          NaN          NaN          NaN          NaN  
1012          NaN          NaN          NaN          NaN  

[5 rows x 51 columns]

When analyzing a dataset, it's always a good idea to start by getting basic information about your dataframe. We can do this by using the `info()` method.

This method can be used to get a short summary of the dataframe.


In [15]:
df_can.info(verbose=False)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1013 entries, 0 to 1012
Columns: 51 entries, Type to Unnamed: 50
dtypes: float64(45), object(6)
memory usage: 403.7+ KB


To get the list of column headers we can call upon the data frame's `columns` instance variable.


In [16]:
df_can.columns

Index([       'Type',    'Coverage',      'OdName',        'AREA',
          'AreaName',         'REG',     'RegName',         'DEV',
           'DevName',          1980,          1981,          1982,
                1983,          1984,          1985,          1986,
                1987,          1988,          1989,          1990,
                1991,          1992,          1993,          1994,
                1995,          1996,          1997,          1998,
                1999,          2000,          2001,          2002,
                2003,          2004,          2005,          2006,
                2007,          2008,          2009,          2010,
                2011,          2012,          2013, 'Unnamed: 43',
       'Unnamed: 44', 'Unnamed: 45', 'Unnamed: 46', 'Unnamed: 47',
       'Unnamed: 48', 'Unnamed: 49', 'Unnamed: 50'],
      dtype='object')

Similarly, to get the list of indices we use the `.index` instance variables.


In [17]:
df_can.index

RangeIndex(start=0, stop=1013, step=1)

Note: The default type of intance variables `index` and `columns` are **NOT** `list`.


In [18]:
print(type(df_can.columns))
print(type(df_can.index))

<class 'pandas.core.indexes.base.Index'>
<class 'pandas.core.indexes.range.RangeIndex'>


To get the index and columns as lists, we can use the `tolist()` method.


In [19]:
df_can.columns.tolist()

['Type',
 'Coverage',
 'OdName',
 'AREA',
 'AreaName',
 'REG',
 'RegName',
 'DEV',
 'DevName',
 1980,
 1981,
 1982,
 1983,
 1984,
 1985,
 1986,
 1987,
 1988,
 1989,
 1990,
 1991,
 1992,
 1993,
 1994,
 1995,
 1996,
 1997,
 1998,
 1999,
 2000,
 2001,
 2002,
 2003,
 2004,
 2005,
 2006,
 2007,
 2008,
 2009,
 2010,
 2011,
 2012,
 2013,
 'Unnamed: 43',
 'Unnamed: 44',
 'Unnamed: 45',
 'Unnamed: 46',
 'Unnamed: 47',
 'Unnamed: 48',
 'Unnamed: 49',
 'Unnamed: 50']

In [20]:
df_can.index.tolist()

[0,
 1,
 2,
 3,
 4,
 5,
 6,
 7,
 8,
 9,
 10,
 11,
 12,
 13,
 14,
 15,
 16,
 17,
 18,
 19,
 20,
 21,
 22,
 23,
 24,
 25,
 26,
 27,
 28,
 29,
 30,
 31,
 32,
 33,
 34,
 35,
 36,
 37,
 38,
 39,
 40,
 41,
 42,
 43,
 44,
 45,
 46,
 47,
 48,
 49,
 50,
 51,
 52,
 53,
 54,
 55,
 56,
 57,
 58,
 59,
 60,
 61,
 62,
 63,
 64,
 65,
 66,
 67,
 68,
 69,
 70,
 71,
 72,
 73,
 74,
 75,
 76,
 77,
 78,
 79,
 80,
 81,
 82,
 83,
 84,
 85,
 86,
 87,
 88,
 89,
 90,
 91,
 92,
 93,
 94,
 95,
 96,
 97,
 98,
 99,
 100,
 101,
 102,
 103,
 104,
 105,
 106,
 107,
 108,
 109,
 110,
 111,
 112,
 113,
 114,
 115,
 116,
 117,
 118,
 119,
 120,
 121,
 122,
 123,
 124,
 125,
 126,
 127,
 128,
 129,
 130,
 131,
 132,
 133,
 134,
 135,
 136,
 137,
 138,
 139,
 140,
 141,
 142,
 143,
 144,
 145,
 146,
 147,
 148,
 149,
 150,
 151,
 152,
 153,
 154,
 155,
 156,
 157,
 158,
 159,
 160,
 161,
 162,
 163,
 164,
 165,
 166,
 167,
 168,
 169,
 170,
 171,
 172,
 173,
 174,
 175,
 176,
 177,
 178,
 179,
 180,
 181,
 182,
 183,
 184,


In [21]:
print(type(df_can.columns.tolist()))
print(type(df_can.index.tolist()))

<class 'list'>
<class 'list'>


To view the dimensions of the dataframe, we use the `shape` instance variable of it.


In [22]:
# size of dataframe (rows, columns)
df_can.shape

(1013, 51)

**Note**: The main types stored in *pandas* objects are `float`, `int`, `bool`, `datetime64[ns]`, `datetime64[ns, tz]`, `timedelta[ns]`, `category`, and `object` (string). In addition, these dtypes have item sizes, e.g. `int64` and `int32`.


Let's clean the data set to remove a few unnecessary columns. We can use *pandas* `drop()` method as follows:


In [23]:
# in pandas axis=0 represents rows (default) and axis=1 represents columns.
df_can.drop(['AREA','REG','DEV','Type','Coverage'], axis=1, inplace=True)
df_can.head(2)

OdName AreaName          RegName             DevName  1980  1981  \
0  Afghanistan     Asia    Southern Asia  Developing regions  16.0  39.0   
1      Albania   Europe  Southern Europe   Developed regions   1.0   0.0   

   1982  1983  1984   1985  ...    2012    2013  Unnamed: 43  Unnamed: 44  \
0  39.0  47.0  71.0  340.0  ...  2635.0  2004.0          NaN          NaN   
1   0.0   0.0   0.0    0.0  ...   620.0   603.0          NaN          NaN   

   Unnamed: 45  Unnamed: 46  Unnamed: 47  Unnamed: 48  Unnamed: 49  \
0          NaN          NaN          NaN          NaN          NaN   
1          NaN          NaN          NaN          NaN          NaN   

   Unnamed: 50  
0          NaN  
1          NaN  

[2 rows x 46 columns]

Let's rename the columns so that they make sense. We can use `rename()` method by passing in a dictionary of old and new names as follows:


In [24]:
df_can.rename(columns={'OdName':'Country', 'AreaName':'Continent', 'RegName':'Region'}, inplace=True)
df_can.columns

Index([    'Country',   'Continent',      'Region',     'DevName',
                1980,          1981,          1982,          1983,
                1984,          1985,          1986,          1987,
                1988,          1989,          1990,          1991,
                1992,          1993,          1994,          1995,
                1996,          1997,          1998,          1999,
                2000,          2001,          2002,          2003,
                2004,          2005,          2006,          2007,
                2008,          2009,          2010,          2011,
                2012,          2013, 'Unnamed: 43', 'Unnamed: 44',
       'Unnamed: 45', 'Unnamed: 46', 'Unnamed: 47', 'Unnamed: 48',
       'Unnamed: 49', 'Unnamed: 50'],
      dtype='object')

We will also add a 'Total' column that sums up the total immigrants by country over the entire period 1980 - 2013, as follows:


In [25]:
df_can['Total'] = df_can.sum(axis=1)
df_can['Total']

0       58639.0
1       15699.0
2       69439.0
3           6.0
4          15.0
         ...   
1008        0.0
1009        0.0
1010        0.0
1011        0.0
1012        0.0
Name: Total, Length: 1013, dtype: float64

We can check to see how many null objects we have in the dataset as follows:


In [26]:
df_can.isnull().sum()

Country         816
Continent       816
Region          816
DevName         816
1980            816
1981            816
1982            816
1983            816
1984            816
1985            816
1986            816
1987            816
1988            816
1989            816
1990            816
1991            816
1992            816
1993            816
1994            816
1995            816
1996            816
1997            816
1998            816
1999            816
2000            816
2001            816
2002            816
2003            816
2004            816
2005            816
2006            816
2007            816
2008            816
2009            816
2010            816
2011            816
2012            816
2013            816
Unnamed: 43    1013
Unnamed: 44    1013
Unnamed: 45    1013
Unnamed: 46    1013
Unnamed: 47    1013
Unnamed: 48    1013
Unnamed: 49    1013
Unnamed: 50    1013
Total             0
dtype: int64

Finally, let's view a quick summary of each column in our dataframe using the `describe()` method.


In [27]:
df_can.describe()

1980           1981           1982          1983  \
count     197.000000     197.000000     197.000000    197.000000   
mean     1453.167513    1306.000000    1230.203046    905.431472   
std     10784.524807    9449.373841    8864.905615   6503.149859   
min         0.000000       0.000000       0.000000      0.000000   
25%         0.000000       0.000000       0.000000      0.000000   
50%        14.000000      10.000000      12.000000     12.000000   
75%       266.000000     299.000000     299.000000    197.000000   
max    143137.000000  128641.000000  121175.000000  89185.000000   

               1984          1985          1986           1987           1988  \
count    197.000000    197.000000    197.000000     197.000000     197.000000   
mean     896.162437    856.304569   1008.639594    1543.908629    1640.456853   
std     6452.570413   6155.858422   7205.630681   11045.558746   11809.252241   
min        0.000000      0.000000      0.000000       0.000000       0.000000   
25%        0.000000      0.000000      1.000000       1.000000       1.000000   
50%       14.000000     17.000000     21.000000      33.000000      36.000000   
75%      207.000000    202.000000    257.000000     467.000000     476.000000   
max    88272.000000  84346.000000  99351.000000  152075.000000  161585.000000   

                1989  ...           2013  Unnamed: 43  Unnamed: 44  \
count     197.000000  ...     197.000000          0.0          0.0   
mean     1944.670051  ...    2629.654822          NaN          NaN   
std     13939.908653  ...   18838.224387          NaN          NaN   
min         0.000000  ...       0.000000          NaN          NaN   
25%         1.000000  ...      45.000000          NaN          NaN   
50%        45.000000  ...     217.000000          NaN          NaN   
75%       538.000000  ...     852.000000          NaN          NaN   
max    191550.000000  ...  259021.000000          NaN          NaN   

       Unnamed: 45  Unnamed: 46  Unnamed: 47  Unnamed: 48  Unnamed: 49  \
count          0.0          0.0          0.0          0.0          0.0   
mean           NaN          NaN          NaN          NaN          NaN   
std            NaN          NaN          NaN          NaN          NaN   
min            NaN          NaN          NaN          NaN          NaN   
25%            NaN          NaN          NaN          NaN          NaN   
50%            NaN          NaN          NaN          NaN          NaN   
75%            NaN          NaN          NaN          NaN          NaN   
max            NaN          NaN          NaN          NaN          NaN   

       Unnamed: 50         Total  
count          0.0  1.013000e+03  
mean           NaN  1.367070e+04  
std            NaN  2.219784e+05  
min            NaN  0.000000e+00  
25%            NaN  0.000000e+00  
50%            NaN  0.000000e+00  
75%            NaN  0.000000e+00  
max            NaN  6.924212e+06  

[8 rows x 43 columns]

---
## *pandas* Intermediate: Indexing and Selection (slicing)<a id="3"></a>


### Select Column
**There are two ways to filter on a column name:**

Method 1: Quick and easy, but only works if the column name does NOT have spaces or special characters.
```python
    df.column_name               # returns series
```

Method 2: More robust, and can filter on multiple columns.

```python
    df['column']                  # returns series
```

```python 
    df[['column 1', 'column 2']]  # returns dataframe
```
---


Example: Let's try filtering on the list of countries ('Country').


In [28]:
df_can.Country  # returns a series

0          Afghanistan
1              Albania
2              Algeria
3       American Samoa
4              Andorra
             ...      
1008               NaN
1009               NaN
1010               NaN
1011               NaN
1012               NaN
Name: Country, Length: 1013, dtype: object

Let's try filtering on the list of countries ('Country') and the data for years: 1980 - 1985.


In [29]:
df_can[['Country', 1980, 1981, 1982, 1983, 1984, 1985]] # returns a dataframe
# notice that 'Country' is string, and the years are integers. 
# for the sake of consistency, we will convert all column names to string later on.

Country  1980  1981  1982  1983  1984   1985
0        Afghanistan  16.0  39.0  39.0  47.0  71.0  340.0
1            Albania   1.0   0.0   0.0   0.0   0.0    0.0
2            Algeria  80.0  67.0  71.0  69.0  63.0   44.0
3     American Samoa   0.0   1.0   0.0   0.0   0.0    0.0
4            Andorra   0.0   0.0   0.0   0.0   0.0    0.0
...              ...   ...   ...   ...   ...   ...    ...
1008             NaN   NaN   NaN   NaN   NaN   NaN    NaN
1009             NaN   NaN   NaN   NaN   NaN   NaN    NaN
1010             NaN   NaN   NaN   NaN   NaN   NaN    NaN
1011             NaN   NaN   NaN   NaN   NaN   NaN    NaN
1012             NaN   NaN   NaN   NaN   NaN   NaN    NaN

[1013 rows x 7 columns]

### Select Row

There are main 2 ways to select rows:

```python
    df.loc[label]    # filters by the labels of the index/column
    df.iloc[index]   # filters by the positions of the index/column
```


Before we proceed, notice that the default index of the dataset is a numeric range from 0 to 194. This makes it very difficult to do a query by a specific country. For example to search for data on Japan, we need to know the corresponding index value.

This can be fixed very easily by setting the 'Country' column as the index using `set_index()` method.


In [30]:
df_can.set_index('Country', inplace=True)
# tip: The opposite of set is reset. So to reset the index, we can use df_can.reset_index()

In [52]:
import pandas as pd

# Assuming df_can is your DataFrame
# Step 1: Drop rows where the value of the index column "Country" is NaN
df_can_cleaned_rows = df_can[df_can.index.notna()]

# Step 2: Drop the specified columns
columns_to_drop = ['Unnamed: 43', 'Unnamed: 44', 'Unnamed: 45', 'Unnamed: 46',
                   'Unnamed: 47', 'Unnamed: 48', 'Unnamed: 49', 'Unnamed: 50']
df_can_cleaned = df_can_cleaned_rows.drop(columns=columns_to_drop)

# Assign the final cleaned DataFrame back to df_can
df_can = df_can_cleaned

print(df_can)


               Continent           Region             DevName      1980  \
Afghanistan         Asia    Southern Asia  Developing regions      16.0   
Albania           Europe  Southern Europe   Developed regions       1.0   
Algeria           Africa  Northern Africa  Developing regions      80.0   
American Samoa   Oceania        Polynesia  Developing regions       0.0   
Andorra           Europe  Southern Europe   Developed regions       0.0   
...                  ...              ...                 ...       ...   
Yemen               Asia     Western Asia  Developing regions       1.0   
Zambia            Africa   Eastern Africa  Developing regions      11.0   
Zimbabwe          Africa   Eastern Africa  Developing regions      72.0   
Unknown            World            World               World   44000.0   
Total              World            World               World  143137.0   

                    1981      1982     1983     1984     1985     1986  ...  \
Afghanistan         

In [54]:
# optional: to remove the name of the index
df_can.index.name = None

Example: Let's view the number of immigrants from Japan (row 87) for the following scenarios:
    1. The full row data (all columns)
    2. For year 2013
    3. For years 1980 to 1985


In [55]:
# 1. the full row data (all columns)
df_can.loc['Japan']

Continent                 Asia
Region            Eastern Asia
DevName      Developed regions
1980                       701
1981                       756
1982                       598
1983                       309
1984                       246
1985                       198
1986                       248
1987                       422
1988                       324
1989                       494
1990                       379
1991                       506
1992                       605
1993                       907
1994                       956
1995                       826
1996                       994
1997                       924
1998                       897
1999                      1083
2000                      1010
2001                      1092
2002                       806
2003                       817
2004                       973
2005                      1067
2006                      1212
2007                      1250
2008                      1284
2009    

In [56]:
# alternate methods
df_can.iloc[87]

Continent                 Asia
Region            Eastern Asia
DevName      Developed regions
1980                       701
1981                       756
1982                       598
1983                       309
1984                       246
1985                       198
1986                       248
1987                       422
1988                       324
1989                       494
1990                       379
1991                       506
1992                       605
1993                       907
1994                       956
1995                       826
1996                       994
1997                       924
1998                       897
1999                      1083
2000                      1010
2001                      1092
2002                       806
2003                       817
2004                       973
2005                      1067
2006                      1212
2007                      1250
2008                      1284
2009    

In [57]:
df_can[df_can.index == 'Japan']

Continent        Region            DevName   1980   1981   1982   1983  \
Japan      Asia  Eastern Asia  Developed regions  701.0  756.0  598.0  309.0   

        1984   1985   1986  ...    2005    2006    2007    2008    2009  \
Japan  246.0  198.0  248.0  ...  1067.0  1212.0  1250.0  1284.0  1194.0   

         2010    2011    2012   2013    Total  
Japan  1168.0  1265.0  1214.0  982.0  27707.0  

[1 rows x 38 columns]

In [58]:
# 2. for year 2013
df_can.loc['Japan', 2013]

982.0

In [59]:
# alternate method
# year 2013 is the last column, with a positional index of 36
df_can.iloc[87, 36]

982.0

In [60]:
# 3. for years 1980 to 1985
df_can.loc['Japan', [1980, 1981, 1982, 1983, 1984, 1984]]

1980    701
1981    756
1982    598
1983    309
1984    246
1984    246
Name: Japan, dtype: object

In [61]:
# Alternative Method
df_can.iloc[87, [3, 4, 5, 6, 7, 8]]

1980    701
1981    756
1982    598
1983    309
1984    246
1985    198
Name: Japan, dtype: object

**Exercise:** Let's view the number of immigrants from **Haiti** for the following scenarios: <br>1. The full row data (all columns) <br>2. For year 2000 <br>3. For years 1990 to 1995


In [62]:
df_can.loc['Haiti']
df_can.loc['Haiti', 2000]
df_can.loc['Haiti', [1990, 1991, 1992, 1993, 1994, 1995]]

1990    2379
1991    2829
1992    2399
1993    3655
1994    2100
1995    2014
Name: Haiti, dtype: object

<details><summary>Click here for a sample python solution</summary>

```python
   # 1. the full row data (all columns)
    df_can.loc['Haiti']
    #or
    df_can[df_can.index == 'Haiti']

    # 2. for year 2000
    df_can.loc['Haiti', 2000]
           
    #  3. for years 1990 to 1995
    df_can.loc['Haiti', [1990, 1991, 1992, 1993, 1994, 1995]]
 ```

</details>


Column names that are integers (such as the years) might introduce some confusion. For example, when we are referencing the year 2013, one might confuse that when the 2013th positional index. 

To avoid this ambuigity, let's convert the column names into strings: '1980' to '2013'.


In [64]:
df_can.columns = list(map(str, df_can.columns))
[print (type(x)) for x in df_can.columns.values] #<-- uncomment to check type of column headers

<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>
<class 'str'>


[None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None]

Since we converted the years to string, let's declare a variable that will allow us to easily call upon the full range of years:


In [65]:
# useful for plotting later on
years = list(map(str, range(1980, 2014)))
years

['1980',
 '1981',
 '1982',
 '1983',
 '1984',
 '1985',
 '1986',
 '1987',
 '1988',
 '1989',
 '1990',
 '1991',
 '1992',
 '1993',
 '1994',
 '1995',
 '1996',
 '1997',
 '1998',
 '1999',
 '2000',
 '2001',
 '2002',
 '2003',
 '2004',
 '2005',
 '2006',
 '2007',
 '2008',
 '2009',
 '2010',
 '2011',
 '2012',
 '2013']

**Exercise:** Create a list named 'year' using map function for years ranging from 1990 to 2013. <br>Then extract the data series from the dataframe df_can for Haiti using year list. 


In [69]:
year = list(map(str, range(1990, 2014)))
haiti_data = df_can.loc['Haiti', year]
haiti_data

1990    2379
1991    2829
1992    2399
1993    3655
1994    2100
1995    2014
1996    1955
1997    1645
1998    1295
1999    1439
2000    1631
2001    2433
2002    2174
2003    1930
2004    1652
2005    1682
2006    1619
2007    1598
2008    2491
2009    2080
2010    4744
2011    6503
2012    5868
2013    4152
Name: Haiti, dtype: object

<details><summary>Click here for a sample python solution</summary>

```python
    #The correct answer is:
    year = list(map(str, range(1990, 2014)))
    haiti = df_can.loc['Haiti', year] # passing in years 1990 - 2013
    
```
</details>


### Filtering based on a criteria <a id="4"></a>
To filter the dataframe based on a condition, we simply pass the condition as a boolean vector. 

For example, Let's filter the dataframe to show the data on Asian countries (AreaName = Asia).


In [70]:
# 1. create the condition boolean series
condition = df_can['Continent'] == 'Asia'
print(condition)

Afghanistan        True
Albania           False
Algeria           False
American Samoa    False
Andorra           False
                  ...  
Yemen              True
Zambia            False
Zimbabwe          False
Unknown           False
Total             False
Name: Continent, Length: 197, dtype: bool


In [71]:
# 2. pass this condition into the dataFrame
df_can[condition]

Continent              Region  \
Afghanistan                                         Asia       Southern Asia   
Armenia                                             Asia        Western Asia   
Azerbaijan                                          Asia        Western Asia   
Bahrain                                             Asia        Western Asia   
Bangladesh                                          Asia       Southern Asia   
Bhutan                                              Asia       Southern Asia   
Brunei Darussalam                                   Asia  South-Eastern Asia   
Cambodia                                            Asia  South-Eastern Asia   
China                                               Asia        Eastern Asia   
China, Hong Kong Special Administrative Region      Asia        Eastern Asia   
China, Macao Special Administrative Region          Asia        Eastern Asia   
Cyprus                                              Asia        Western Asia   
Democratic People's Republic of Korea               Asia        Eastern Asia   
Georgia                                             Asia        Western Asia   
India                                               Asia       Southern Asia   
Indonesia                                           Asia  South-Eastern Asia   
Iran (Islamic Republic of)                          Asia       Southern Asia   
Iraq                                                Asia        Western Asia   
Israel                                              Asia        Western Asia   
Japan                                               Asia        Eastern Asia   
Jordan                                              Asia        Western Asia   
Kazakhstan                                          Asia        Central Asia   
Kuwait                                              Asia        Western Asia   
Kyrgyzstan                                          Asia        Central Asia   
Lao People's Democratic Republic                    Asia  South-Eastern Asia   
Lebanon                                             Asia        Western Asia   
Malaysia                                            Asia  South-Eastern Asia   
Maldives                                            Asia       Southern Asia   
Mongolia                                            Asia        Eastern Asia   
Myanmar                                             Asia  South-Eastern Asia   
Nepal                                               Asia       Southern Asia   
Oman                                                Asia        Western Asia   
Pakistan                                            Asia       Southern Asia   
Philippines                                         Asia  South-Eastern Asia   
Qatar                                               Asia        Western Asia   
Republic of Korea                                   Asia        Eastern Asia   
Saudi Arabia                                        Asia        Western Asia   
Singapore                                           Asia  South-Eastern Asia   
Sri Lanka                                           Asia       Southern Asia   
State of Palestine                                  Asia        Western Asia   
Syrian Arab Republic                                Asia        Western Asia   
Tajikistan                                          Asia        Central Asia   
Thailand                                            Asia  South-Eastern Asia   
Turkey                                              Asia        Western Asia   
Turkmenistan                                        Asia        Central Asia   
United Arab Emirates                                Asia        Western Asia   
Uzbekistan                                          Asia        Central Asia   
Viet Nam                                            Asia  South-Eastern Asia   
Yemen                                               Asia        Western Asia   

                                              

In [73]:
# we can pass multiple criteria in the same line.
# let's filter for AreaNAme = Asia and RegName = Southern Asia

df_can[(df_can['Continent']=='Asia') & (df_can['Region']=='Southern Asia')]

# note: When using 'and' and 'or' operators, pandas requires we use '&' and '|' instead of 'and' and 'or'
# don't forget to enclose the two conditions in parentheses

Continent         Region             DevName  \
Afghanistan                     Asia  Southern Asia  Developing regions   
Bangladesh                      Asia  Southern Asia  Developing regions   
Bhutan                          Asia  Southern Asia  Developing regions   
India                           Asia  Southern Asia  Developing regions   
Iran (Islamic Republic of)      Asia  Southern Asia  Developing regions   
Maldives                        Asia  Southern Asia  Developing regions   
Nepal                           Asia  Southern Asia  Developing regions   
Pakistan                        Asia  Southern Asia  Developing regions   
Sri Lanka                       Asia  Southern Asia  Developing regions   

                              1980    1981    1982    1983    1984    1985  \
Afghanistan                   16.0    39.0    39.0    47.0    71.0   340.0   
Bangladesh                    83.0    84.0    86.0    81.0    98.0    92.0   
Bhutan                         0.0     0.0     0.0     0.0     1.0     0.0   
India                       8880.0  8670.0  8147.0  7338.0  5704.0  4211.0   
Iran (Islamic Republic of)  1172.0  1429.0  1822.0  1592.0  1977.0  1648.0   
Maldives                       0.0     0.0     0.0     1.0     0.0     0.0   
Nepal                          1.0     1.0     6.0     1.0     2.0     4.0   
Pakistan                     978.0   972.0  1201.0   900.0   668.0   514.0   
Sri Lanka                    185.0   371.0   290.0   197.0  1086.0   845.0   

                              1986  ...     2005     2006     2007     2008  \
Afghanistan                  496.0  ...   3436.0   3009.0   2652.0   2111.0   
Bangladesh                   486.0  ...   4171.0   4014.0   2897.0   2939.0   
Bhutan                         0.0  ...      5.0     10.0      7.0     36.0   
India                       7150.0  ...  36210.0  33848.0  28742.0  28261.0   
Iran (Islamic Republic of)  1794.0  ...   5837.0   7480.0   6974.0   6475.0   
Maldives                       0.0  ...      0.0      0.0      2.0      1.0   
Nepal                         13.0  ...    607.0    540.0    511.0    581.0   
Pakistan                     691.0  ...  14314.0  13127.0  10124.0   8994.0   
Sri Lanka                   1838.0  ...   4930.0   4714.0   4123.0   4756.0   

                               2009     2010     2011     2012     2013  \
Afghanistan                  1746.0   1758.0   2203.0   2635.0   2004.0   
Bangladesh                   2104.0   4721.0   2694.0   2640.0   3789.0   
Bhutan                        865.0   1464.0   1879.0   1075.0    487.0   
India                       29456.0  34235.0  27509.0  30933.0  33087.0   
Iran (Islamic Republic of)   6580.0   7477.0   7479.0   7534.0  11291.0   
Maldives                        7.0      4.0      3.0      1.0      1.0   
Nepal                         561.0   1392.0   1129.0   1185.0   1308.0   
Pakistan                     7217.0   6811.0   7468.0  11227.0  12603.0   
Sri Lanka                    4547.0   4422.0   3309.0   3338.0   2394.0   

                               Total  
Afghanistan                  58639.0  
Bangladesh                   65568.0  
Bhutan                        5876.0  
India                       691904.0  
Iran (Islamic Republic of)  175923.0  
Maldives                        30.0  
Nepal                        10222.0  
Pakistan                    241600.0  
Sri Lanka                   148358.0  

[9 rows x 38 columns]

**Exercise:** Fetch the data where AreaName is 'Africa' and RegName is 'Southern Africa'. <br>Display the dataframe and find out how many instances are there?


In [75]:
df_can[(df_can["Continent"] == "Africa") & (df_can["Region"] == "Southern Africa")]

Continent           Region             DevName    1980    1981  \
Botswana        Africa  Southern Africa  Developing regions    10.0     1.0   
Lesotho         Africa  Southern Africa  Developing regions     1.0     1.0   
Namibia         Africa  Southern Africa  Developing regions     0.0     5.0   
South Africa    Africa  Southern Africa  Developing regions  1026.0  1118.0   
Swaziland       Africa  Southern Africa  Developing regions     4.0     1.0   

               1982   1983   1984   1985   1986  ...   2005    2006    2007  \
Botswana        3.0    3.0    7.0    4.0    2.0  ...    7.0    11.0     8.0   
Lesotho         1.0    2.0    7.0    5.0    3.0  ...    4.0     0.0     4.0   
Namibia         5.0    3.0    2.0    1.0    1.0  ...    6.0    19.0    13.0   
South Africa  781.0  379.0  271.0  310.0  718.0  ...  988.0  1111.0  1200.0   
Swaziland       1.0    0.0   10.0    7.0    1.0  ...    7.0     7.0     5.0   

                2008    2009    2010   2011    2012    2013    Total  
Botswana        28.0    15.0    42.0   53.0    64.0    76.0    396.0  
Lesotho          1.0     8.0     7.0    1.0     0.0     6.0    107.0  
Namibia         26.0    14.0    16.0   23.0    24.0    83.0    320.0  
South Africa  1123.0  1188.0  1238.0  959.0  1243.0  1240.0  40568.0  
Swaziland        6.0    10.0     3.0   13.0    17.0    39.0    188.0  

[5 rows x 38 columns]

<details><summary>Click here for a sample python solution</summary>

```python
    df_can[(df_can['Continent']=='Africa') & (df_can['Region']=='Southern Africa')]
```
</details>


### Sorting Values of a Dataframe or Series <a id="5"></a><br>
You can use the `sort_values()` function is used to sort a DataFrame or a Series based on one or more columns. <br>You to specify the column(s) by which you want to sort and the order (ascending or descending). Below is the syntax to use it:-<br><br>
```df.sort_values(col_name, axis=0, ascending=True, inplace=False, ignore_index=False)```<br><br>
col_nam - the column(s) to sort by. <br>
axis - axis along which to sort. 0 for sorting by rows (default) and 1 for sorting by columns.<br>
ascending - to sort in ascending order (True, default) or descending order (False).<br>
inplace - to perform the sorting operation in-place (True) or return a sorted copy (False, default).<br>
ignore_index - to reset the index after sorting (True) or keep the original index values (False, default).<br>


Let's sort out dataframe df_can on 'Total' column, in descending order to find out the top 5 countries that contributed the most to immigration to Canada. 


In [76]:
df_can.sort_values(by='Total', ascending=False, axis=0, inplace=True)
top_5 = df_can.head(5)
top_5

Continent           Region  \
Total                                                  World            World   
India                                                   Asia    Southern Asia   
China                                                   Asia     Eastern Asia   
United Kingdom of Great Britain and Northern Ir...    Europe  Northern Europe   
Unknown                                                World            World   

                                                               DevName  \
Total                                                            World   
India                                               Developing regions   
China                                               Developing regions   
United Kingdom of Great Britain and Northern Ir...   Developed regions   
Unknown                                                          World   

                                                        1980      1981  \
Total                                               143137.0  128641.0   
India                                                 8880.0    8670.0   
China                                                 5123.0    6682.0   
United Kingdom of Great Britain and Northern Ir...   22045.0   24796.0   
Unknown                                              44000.0   18078.0   

                                                        1982     1983  \
Total                                               121175.0  89185.0   
India                                                 8147.0   7338.0   
China                                                 3308.0   1863.0   
United Kingdom of Great Britain and Northern Ir...   20620.0  10015.0   
Unknown                                              16904.0  13635.0   

                                                       1984     1985     1986  \
Total                                               88272.0  84346.0  99351.0   
India                                                5704.0   4211.0   7150.0   
China                                                1527.0   1816.0   1960.0   
United Kingdom of Great Britain and Northern Ir...  10170.0   9564.0   9470.0   
Unknown                                             14855.0  14368.0  13303.0   

                                                    ...      2005      2006  \
Total                                               ...  262242.0  251640.0   
India                                               ...   36210.0   33848.0   
China                                               ...   42584.0   33518.0   
United Kingdom of Great Britain and Northern Ir...  ...    7258.0    7140.0   
Unknown                                             ...    4785.0    4583.0   

                                                        2007      2008  \
Total                                               236753.0  247244.0   
India                                                28742.0   28261.0   
China                                                27642.0   30037.0   
United Kingdom of Great Britain and Northern Ir...    8216.0    8979.0   
Unknown                                               4348.0    4197.0   

                                                        2009      2010  \
Total                                               252170.0  280687.0   
India                                                29456.0   34235.0   
China                                                29622.0   30391.0   
United Kingdom of Great Britain and Northern Ir...    8876.0    8724.0   
Unknown                                               3402.0    3731.0   

                                                        2011      2012  \
Total                                               248748.0  257903.0   
India                                                27509.0   30933.0   
China                                                28502.0   33024.0   
United Kingdom of Great Britain and Northern Ir...    6204.0    6195.0   
Unknown      

**Exercise:** Find out top 3 countries that contributes the most to immigration to Canda in the year 2010. <br> Display the country names with the immigrant count in this year


In [78]:
df_can.sort_values(by='2010', ascending=False, axis=0, inplace=True)
top3_2010 = df_can['2010'].head(3)
top3_2010

Total          280687.0
Philippines     38617.0
India           34235.0
Name: 2010, dtype: float64

Congratulations! you have learned how to wrangle data with Pandas. You will be using alot of these commands to preprocess the data before its can be used for data visualization.


### Thank you for completing this lab!


## Author

<a href="https://www.linkedin.com/in/aklson/" target="_blank">Alex Aklson</a>


### Other Contributors
[Jay Rajasekharan](https://www.linkedin.com/in/jayrajasekharan),
[Ehsan M. Kermani](https://www.linkedin.com/in/ehsanmkermani),
[Slobodan Markovic](https://www.linkedin.com/in/slobodan-markovic),
[Weiqing Wang](https://www.linkedin.com/in/weiqing-wang-641640133/),
[Dr. Pooja](https://www.linkedin.com/in/p-b28802262/)


## Change Log


|  Date (YYYY-MM-DD) | Version | Changed By    |  Change Description                   |
|--------------------|---------|---------------|---------------------------------------|
| 2023-06-08         | 2.5     | Dr. Pooja         |  Separated from original lab        |
| 2021-05-29         | 2.4     | Weiqing Wang  |  Fixed typos and code smells.         |
| 2021-01-20         | 2.3     | Lakshmi Holla |  Changed TOC cell markdown            |
| 2020-11-20         | 2.2     | Lakshmi Holla |  Changed IBM box URL                  |
| 2020-11-03         | 2.1     | Lakshmi Holla |  Changed URL and info method          |
| 2020-08-27         | 2.0     | Lavanya       |  Moved Lab to course repo in GitLab   |




## <h3 align="center"> © IBM Corporation 2020. All rights reserved. <h3/>
